### PR TITLE
External image controller

### PR DESCRIPTION
**Pull request by soltysh**
This switches image controller to use external images API.
/assign @bparees @mfojtik

### CODE REVIEW

**Comment by soltysh** [comment]
@bparees @mfojtik this is ready for reviews

File: pkg/image/util/helpers.go [hunk 1]
```diff
@@ -3,18 +3,23 @@ package util
...
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/sets"
 
 	dockerv10 "github.com/openshift/api/image/docker10"
 	imagev1 "github.com/openshift/api/image/v1"
 	imageapi "github.com/openshift/origin/pkg/image/apis/image"
 	dockerapi10 "github.com/openshift/origin/pkg/image/apis/image/docker10"
+	imagereference "github.com/openshift/origin/pkg/image/apis/image/reference"
+	"github.com/openshift/origin/pkg/image/internal/digest"
```
Comment (mfojtik) [inline pkg/image/util/helpers.go]: there seems to be nothing internal in this package, I would move it to pkg/image/util

File: pkg/cmd/openshift-controller-manager/controller/interfaces.go [hunk 2]
```diff
@@ -191,9 +190,9 @@ func (c *ControllerContext) StartInformers(stopCh <-chan struct{}) {
 
 	c.AppsInformers.Start(stopCh)
 	c.BuildInformers.Start(stopCh)
+	c.ImageInformers.Start(stopCh)
 	c.NetworkInformers.Start(stopCh)
 
-	c.InternalImageInformers.Start(stopCh)
```
Comment (mfojtik) [inline pkg/cmd/openshift-controller-manager/controller/interfaces.go]: 👍

File: pkg/image/controller/imagestream_controller.go [hunk 3]
```diff
@@ -238,38 +241,47 @@ func handleImageStream(
...
-	for tag, tagRef := range stream.Spec.Tags {
+	for _, tagRef := range stream.Spec.Tags {
 		if tagImportable(tagRef) &&
-			(tagNeedsImport(stream, tag, tagRef, true) || !partial) {
-			isi.Spec.Images = append(isi.Spec.Images, imageapi.ImageImportSpec{
-				From:            kapi.ObjectReference{Kind: "DockerImage", Name: tagRef.From.Name},
-				To:              &kapi.LocalObjectReference{Name: tag},
+			(tagNeedsImport(stream, tagRef, true) || !partial) {
+			isi.Spec.Images = append(isi.Spec.Images, imagev1.ImageImportSpec{
+				From:            corev1.ObjectReference{Kind: "DockerImage", Name: tagRef.From.Name},
```
Comment (mfojtik) [inline pkg/image/controller/imagestream_controller.go]: do we need APIVersion here?

File: pkg/image/controller/trigger/image_trigger_controller_test.go [hunk 4]
```diff
@@ -262,17 +263,20 @@ func TestTriggerControllerSyncImageStream(t *testing.T) {
...
 func TestTriggerControllerSyncBuildConfigResource(t *testing.T) {
 	tests := []struct {
-		name    string
-		is      *imageapi.ImageStream
-		bc      *buildv1.BuildConfig
-		tagResp []fakeTagResponse
-		req     *buildv1.BuildRequest
+		name string
+		is   *imagev1.ImageStream
+		// TODO: get rid of this once buildgenerator is on externals
```
Comment (mfojtik) [inline pkg/image/controller/trigger/image_trigger_controller_test.go]: https://github.com/openshift/origin/pull/20772

File: pkg/image/util/helpers.go (hunk 1 above)
Comment (soltysh) [reply pkg/image/util/helpers.go]: I think this `internal` package was to have a stable digest and reference implementation, not to rely on docker ones. But I'd defer to @dmage or @legionus they will know better. I just moved it to a place where the new helpers can use it.

**Comment by mfojtik** [comment]
@soltysh refactoring looks good, feel free to retag this for merge once you sort out the quota tests.

/lgtm

File: pkg/image/controller/imagestream_controller.go (hunk 3 above)
Comment (soltysh) [reply pkg/image/controller/imagestream_controller.go]: For now we can get away without it, but we could add it to be fully API-compliant.

File: pkg/quota/image/registry.go [hunk 5]
```diff
@@ -8,18 +8,19 @@ import (
...
+	imagev1informer "github.com/openshift/client-go/image/informers/externalversions/image/v1"
 	"github.com/openshift/origin/pkg/api/legacy"
 	imageapi "github.com/openshift/origin/pkg/image/apis/image"
-	imageinternalversion "github.com/openshift/origin/pkg/image/generated/informers/internalversion/image/internalversion"
-	imageclient "github.com/openshift/origin/pkg/image/generated/internalclientset/typed/image/internalversion"
 )
 
 var legacyObjectCountAliases = map[schema.GroupVersionResource]kapi.ResourceName{
-	imageapi.GroupVersion.WithResource("imagestreams"): imageapi.ResourceImageStreams,
+	imagev1.GroupVersion.WithResource("imagestreams"): imageapi.ResourceImageStreams,
```
Comment (soltysh) [inline pkg/quota/image/registry.go]: @deads2k @mfojtik this and a similar change bellow was missing causing quota tests to fail.

**Comment by soltysh** [comment]
Rebased, re-applying the label back.

**Comment by smarterclayton** [comment]
I think this broke integration tests: 

https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/20817/pull-ci-origin-integration/2559/

Not sure how it merged if so, but this is the only major change I can see to move from internal build types to external build types (unless there was another PR...)

**Comment by smarterclayton** [comment]
```
=== RUN   TestIntegration/TestSimpleImageChangeBuildTriggerFromImageStreamTagSTIWithConfigChange
=== PAUSE TestIntegration/TestSimpleImageChangeBuildTriggerFromImageStreamTagSTIWithConfigChange
=== CONT  TestIntegration/TestSimpleImageChangeBuildTriggerFromImageStreamTagSTIWithConfigChange
    --- FAIL: TestIntegration/TestSimpleImageChangeBuildTriggerFromImageStreamTagSTIWithConfigChange (107.97s)
    	runner_test.go:129: FAILED TestSimpleImageChangeBuildTriggerFromImageStreamTagSTIWithConfigChange, retrying:
    		
    		panic: interface conversion: runtime.Object is *v1.Build, not *build.Build [recovered]
    			panic: interface conversion: runtime.Object is *v1.Build, not *build.Build
    		
    		goroutine 122 [running]:
    		testing.tRunner.func1(0xc4209acb40)
    			/usr/local/go/src/testing/testing.go:742 +0x29d
    		panic(0x4b1bca0, 0xc43aba4000)
    			/usr/local/go/src/runtime/panic.go:502 +0x229
    		github.com/openshift/origin/test/integration.runTest(0xc4209acb40, 0x55ca53e, 0x32, 0xc428a976c0, 0xc437c0e6e0, 0xc434234000, 0xc428c0a000, 0x54c837d, 0x6)
    			/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/integration/imagechange_buildtrigger_test.go:300 +0x369e
    		github.com/openshift/origin/test/integration.TestSimpleImageChangeBuildTriggerFromImageStreamTagSTIWithConfigChange(0xc4209acb40)
    			/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/integration/imagechange_buildtrigger_test.go:52 +0x4d8
    		testing.tRunner(0xc4209acb40, 0x57a84a0)
    			/usr/local/go/src/testing/testing.go:777 +0xd0
    		created by testing.(*T).Run
    			/usr/local/go/src/testing/testing.go:824 +0x2e0
    		
    		=== OUTPUT
```

**Comment by smarterclayton** [comment]
This started breaking about 3 days ago, when this merged.  It *could* be related to 
https://github.com/openshift/origin/pull/20659 but I don't see earlier failures.

**Comment by bparees** [comment]
@smarterclayton is it an intermittent failure? or did you revert something?  seems like something merged again today.

**Comment by smarterclayton** [comment]
Intermittent.  It could also have been a failure only in Michal’s PR

**Comment by soltysh** [comment]
> Intermittent. It could also have been a failure only in Michal’s PR

I'm guessing this was just that, but I'll keep my eyes opened.

**Pull request merged**